hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/transaction/TestTransactionManager.java

```diff
@@ -0,0 +1,197 @@
+      transactionManager.beginTransaction();
+      transactionManager.endTransaction();
+    }
+  }
+
+  @Test
+  public void testMultiWriterTransactions() {
+    final int threadCount = 3;
+    final long awaitMaxTimeoutMs = 2000L;
+    final CountDownLatch latch = new CountDownLatch(threadCount);
+    final AtomicBoolean writer1Completed = new AtomicBoolean(false);
+    final AtomicBoolean writer2Completed = new AtomicBoolean(false);
+
+    // Let writer1 get the lock first, then wait for others
+    // to join the sync up point.
+    Thread writer1 = new Thread(() -> {
+      assertDoesNotThrow(() -> {
+        transactionManager.beginTransaction();
+      });
+      latch.countDown();
+      try {
+        latch.await(awaitMaxTimeoutMs, TimeUnit.MILLISECONDS);
+        // Following sleep is to make sure writer2 attempts
+        // to try lock and to get bocked on the lock which
+        // this thread is currently holding.
+        Thread.sleep(50);
+      } catch (InterruptedException e) {
+        //
+      }
+      assertDoesNotThrow(() -> {
+        transactionManager.endTransaction();
+      });
+      writer1Completed.set(true);
+    });
+    writer1.start();
+
+    // Writer2 will block on trying to acquire the lock
+    // and will eventually get the lock before the timeout.
+    Thread writer2 = new Thread(() -> {
+      latch.countDown();
+      assertDoesNotThrow(() -> {
+        transactionManager.beginTransaction();
+      });
+      assertDoesNotThrow(() -> {
+        transactionManager.endTransaction();
+      });
+      writer2Completed.set(true);
+    });
+    writer2.start();
+
+    // Let writer1 and writer2 wait at the sync up
+    // point to make sure they run in parallel and
+    // one get blocked by the other.
+    latch.countDown();
+    try {
+      writer1.join();
+      writer2.join();
+    } catch (InterruptedException e) {
+      //
+    }
+
+    // Make sure both writers actually completed good
+    Assertions.assertTrue(writer1Completed.get());
+    Assertions.assertTrue(writer2Completed.get());
+  }
+
+  @Test
+  public void testMultiWriterMultipleTransactions() {
+    final int threadCount = 3;
+    final CountDownLatch latch = new CountDownLatch(threadCount);
+    final AtomicBoolean writer1Completed = new AtomicBoolean(false);
+    final AtomicBoolean writer2Completed = new AtomicBoolean(false);
+
+    Thread writer1 = new Thread(() -> {
+      latch.countDown();
+      for (int i = 0; i < 32; i++) {
+        transactionManager.beginTransaction();
+        transactionManager.endTransaction();
+      }
+      writer1Completed.set(true);
+    });
+    writer1.start();
+
+    Thread writer2 = new Thread(() -> {
+      latch.countDown();
+      for (int i = 0; i < 32; i++) {
+        transactionManager.beginTransaction();
+        transactionManager.endTransaction();
+      }
+      writer2Completed.set(true);
+    });
+    writer2.start();
+
+    // Let writer1 and writer2 wait at the sync up
```

Comment:
did you miss to add the latch.await above? I see only countdowns.

Reply:
fixed.